Public/Src/Utilities/Utilities/Serialization/InliningReader.cs

```diff
@@ -100,35 +101,45 @@ public override PathAtom ReadPathAtom()
 
         /// <inheritdoc />
         public override StringId ReadStringId()
+        {
+            return ReadStringId(kind: InlinedStringKind.Default);
+        }
+
+        /// <todoc />
+        public StringId ReadStringId(InlinedStringKind kind)
         {
             // Read the index
             var index = ReadInt32Compact();
 
             // Check if string already encountered
             if (index > m_maxReadStringIndex)
             {
-                // This is a new string
-                // Read if string is ascii or UTF-16
-                bool isAscii = ReadBoolean();
-
-                // Read the byte length
-                int byteLength = ReadInt32Compact();
-
-                CollectionUtilities.GrowArrayIfNecessary(ref m_buffer, byteLength);
-
-                // Read the bytes into the buffer
-                Read(m_buffer, 0, byteLength);
-
-                var binaryString = new BinaryStringSegment(m_buffer, 0, byteLength, isAscii);
+                var binaryString = ReadStringIdValue(kind, ref m_buffer);
                 var stringId = m_pathTable.StringTable.AddString(binaryString);
-
                 m_readStrings[(uint)index] = stringId;
-
                 m_maxReadStringIndex = index;
-                return stringId;
             }
 
             return m_readStrings[(uint)index];
         }
+
+        /// <todoc />
+        public virtual BinaryStringSegment ReadStringIdValue(InlinedStringKind kind, ref byte[] buffer)

```

Comment:
>kind [](start = 79, length = 4)

Where do we use this `kind`?

Reply:
StringIds can represent more than just strings, and non-strings (paths) need to be remapped, so we have to keep track of what type the stringid represents.

---
In reply to: [289579905](https://github.com/microsoft/BuildXL/pull/394#discussion_r289579905) [](ancestors = 289579905)